### ISSUE TITLE
Custom validaton messages are ignored through the API

### ISSUE THREAD

**Issue by unleashit** [directus/directus#18120]
### Describe the Bug

When validations are set at the field level, they are respected in both the web app and the API. But custom validation messages aren't passed through the API (they do show in the web app). Instead, you always get a default message which includes the full assertion (including any regex or potentially sensitive tests). 

I couldn't find much detail in the docs but I found there are two places to configure validations, at the field level, and by role/crud. It seems the field is the default and the roles are additive which is nice, but the latter method doesn't offer a way to add a custom validation message, which means AFAICT there isn't a way (at least through the GUI) to send custom validation error messages through the API.

As an aside, when there are multiple assertions (with and/or) in a field validation, there isn't a way (AFAICT) to set custom validation messages per error. This is probably more of a feature request though.

### To Reproduce

1. Create a collection with a basic text field.
2. Add any validation (ex: must contain "magicword")
3. Add a custom validation message
4. Grant a role create and/or update perms on at least the field 
5. curl or ajax request to create new item in the collection, including the field but without "magicword" in it.
6. Receive a response with a generic (no custom) error message like:

```javascript
{
	"message": "\"test\" must contain [magicword]",
	"extensions": {
		"code": "FAILED_VALIDATION",
		"field": "test",
		"type": "contains",
		"substring": "test"
	}
}
```

### Hosting Strategy

Self-Hosted (Docker Image)

**Comment by unleashit** [comment]
OK, while I meant the ability to add custom messages "per error" as a separate issue to the main one,  it makes sense that translations should be involved with anything involving text. Will go ahead and do that it in a bit thanks!

**Comment by unleashit** [comment]
Hmm, but I still see this as a bug since too many details are being passed in the response. IMO at a minimum,. a very bare bones message should be sent like "[field] is invalid".

**Comment by unleashit** [comment]
For example I have a text area fields with a regex that tests for a min (and max) number of words. I'd rather not have this regex be returned in the error since it could potentially allow (for example a spam bot on a public collection) to easily guess the criteria. This is just an example and while security via obfuscation isn't perfect, on a different app it did stop a particular bot that would routinely submit forms with only a few words in fields where didn't make sense. Obviously the bot (or human behind it) would have to be custom enough to learn from the response, but hey we're in the ChatGPT era now. Am I wrong to think that you don't always want to expose your exact validation rules? Stuff like "[field] is required" are probably fine.

**Comment by rijkvanzanten** [comment]
Hmm tricky! That argument goes both ways, cause for users who aren't in the know of what's required, there's no way to otherwise know what will pass or fail the validation! I don't think that's a situation we can assume for all use cases, so that would have to be implemented in a configuration option somewhere 🤔 (eg a new feature after all! :D)

**Comment by rijkvanzanten** [comment]
> I can't say I agree with your take, and to be honest find it a bit discouraging. I respect your level of commitment, but I would say in the scheme of things this type of issue is actually much more important than the bigger fun stuff like websockets. [...] I feel this is a bug and not a feature request, because the way the existing feature is implemented has introduced a bug: exposing any and all validation assertions, including those that might be sensitive

This is working as intended and implemented, hence I'm not considering it a bug. Similarly, since this has been implemented, nobody has flagged it as problematic, not on the PR, not in new issues since launch. Even though I personally agree it'd be a nice improvement to configure the strictness of validation output somewhere, I can't just drop everything and prioritize that based on anecdotal evidence of a single person..

Also regarding:

> in the scheme of things this type of issue is actually much more important than the bigger fun stuff like websockets

specifically; what an odd comment to make 🤔 There's a team of core maintainers working on a dozen things in parallel at all times. Nobody suggested issues aren't as/more important than features 🤷🏻 Not a great look to just imply like that we should drop everything else to focus on your particular suggestion first 👀 

> It's interesting that this doesn't seem important. I can't think of a tool or library that doesn't have it. And I've worked with a lot, both server and client side. We're talking everything from other admin GUIs to CMS form builder plugins to JS form libraries to raw schema validators. If it validates, it provides an out of the box way to show a custom error. Have you found otherwise?

Citation needed.. Without any examples you're not really leaving me any room to work with this. Without examples, there's also no way to confirm whether or not those errors are returned in UIs or APIs, as those can be pretty different as well (as is the case with Directus). There's no standardized agreed spec on how to return validation errors in REST/GraphQL APIs as far as I'm aware. As a matter of fact, now that I'm thinking about it, GraphQLs JS reference implementation is pretty open in how it returns validation errors itself (listing things like exact path to the problem in the query) 🤷🏻 

> There are two reasons why generic validation msgs, with the full assertion and without a custom option are bad:
> 1. they are user unfriendly. Echoing out, for example, an entire regex isn't professional and will confuse most people. A friendly message is needed.
> 2. in some cases, they give out too much information. I already gave you two examples above (expecting a secret word, and where you wouldn't want to make it too easy for a bot to discover your character or word minimum). In both of those cases, I would prefer to send a message like, case 1: "You didn't say the magic word, you're not welcome to the club!" or case 2: "Message too short" (or too long, etc.).

Great starting point for a feature request! Re: 1 specifically: Directus includes a way to set a custom validation message for use in the presentation layer (eg the admin app). How you present errors on your own output UIs is up to you to decide.

> I'm a bit surprised I couldn't find an already open (or closed) issue for this. My guess for why not is:
> 1. I imagine a good share of Directus users are pretty newbie and don't think much about server side validation messages because they're using a client forms library. I bet a good portion don't even test their forms without cilent validations to see what happens.

I'd be very careful with those assumptions.. Directus is used from beginner dev hobby projects to giant Fortune 50 corporations

> 2. Wouldn't be surprised if this happens some of the bigger COs too. But if you haven't heard from them, maybe they're just investing the time to fix it.

Same thing there.. we're in close touch with a lot of our very large use cases discussing issues small and large. I don't think it's fair to say "If a bug/feature _hasn't_ been reported it's because people are working around it without saying anything". 

In all honesty, I think the most likely reason why in 11 years of issues on this repo you can't find an issue on this is because people haven't before seen it as a problem..

**Comment by unleashit** [comment]
>  There's a team of core maintainers working on a dozen things in parallel at all times. Nobody suggested issues aren't as/more important than features 🤷🏻 Not a great look to just imply like that we should drop everything else to focus on your particular suggestion first 👀

Sorry you're taking it that way. Of course I don't think you should drop everything over this or any particular thing I meant only that I was a bit disappointed in the way you so quickly dismissed that this is an issue at all. You're basically saying that you think I'm having a unique problem... and you don't see it as an issue at all. Agreeing there's an issue is not the same thing as expecting you to fix it.

> I'd be very careful with those assumptions.. Directus is used from beginner dev hobby projects to giant Fortune 50 corporations

> I don't think it's fair to say "If a bug/feature hasn't been reported it's because people are working around it without saying anything".

I didn't say assume or assume any of this. I just thought it was curious "why" they haven't, and posited a couple of theories. I have no doubt there are a ton of good devs using Directus. Just that I've literally never even encountered a tool that provides custom validation on user input that doesn't provide a way to customize the message. That isn't a put down that you don't provide the said feature, it is simply a statement that I've never encountered a situation where a tool I'm using accepts custom validating logic but doesn't allow you to customize the message. 

Citation needed..

Well sorry if I didn't, I've spent a lot of time on this and don't mean to take up a lot more of either of ours. I figured you'd know what I mean, but maybe(?) it could you've worked on tools more than you've built products? An example would be a client/server schema validator like Yup/Zod/Joi. Any JS form builder (forReact: React Hook Form, Formik, Redux Form).  Drupal, Joomla CCKs (content type builders like what Directus does). I believe out of the box, WP Advanced Custom Fields does too, but I can't remember at this point. That's a short list. All of these products (either code or gui) provide the ability to choose a cust error message. 

I'm sorry if you're not happy with what I have to say, it's not my intention to be negative. At this point, I've accepted that we're probably not going to agree and so it goes. There are bigger problems. At this point, I'm not sure if I'll use Directus again, but not just because of this. I'm really looking for something developer focused, schema as code, more deployment options, etc. Really like the Directus backend and DB, but obviously you can't be everything for everyone.
